File: 02_Data_Structures/advanced_structures/bloom_filter/implementations/bloom_filter.py
"""
Bloom Filter Implementation in Python

Problem Statement:
------------------
Implement a Bloom Filter – a space-efficient probabilistic data structure used
to test whether an element is in a set. False positives are possible, but
false negatives are not.

Use Cases:
----------
- Databases (checking cache or storage quickly)
- Web caching (checking if a URL exists)
- Networking (packet filtering, routing tables)
- Security (malware/URL blacklists)

Core Operations:
----------------
- add(item): Insert an item into the Bloom Filter.
- might_contain(item): Check if an item might be in the Bloom Filter.

Key Properties:
---------------
- False positives possible (element reported present when not).
- False negatives impossible (never misses actual inserted elements).

Complexity:
-----------
- Insertion: O(k)   [k = number of hash functions]
- Query: O(k)
- Space: O(m) bits  [m = size of bit array]
"""

import mmh3  # MurmurHash3 for multiple hash functions (fast, non-cryptographic)
from bitarray import bitarray


class BloomFilter:
    def __init__(self, size: int, num_hashes: int):
        """
        Initialize the Bloom Filter.

        :param size: Size of the bit array (m)
        :param num_hashes: Number of hash functions (k)
        """
        self.size = size
        self.num_hashes = num_hashes
        self.bit_array = bitarray(size)
        self.bit_array.setall(0)

    def _hashes(self, item: str):
        """
        Generate multiple hash values for the given item.

        Uses MurmurHash3 with different seeds to simulate multiple hash functions.

        :param item: String item to hash
        :return: Generator of hash indices
        """
        for seed in range(self.num_hashes):
            yield mmh3.hash(item, seed) % self.size

    def add(self, item: str):
        """
        Insert an item into the Bloom Filter.

        :param item: String item to insert
        """
        for index in self._hashes(item):
            self.bit_array[index] = 1

    def might_contain(self, item: str) -> bool:
        """
        Check if the item might be in the Bloom Filter.

        :param item: String item to check
        :return: True if possibly in set, False if definitely not
        """
        return all(self.bit_array[index] for index in self._hashes(item))

    def __str__(self):
        """Return the bit array as a string for debugging."""
        return self.bit_array.to01()


# Example usage
if __name__ == "__main__":
    # Bloom Filter with 1000 bits and 4 hash functions
    bf = BloomFilter(size=1000, num_hashes=4)

    # Insert some items
    bf.add("apple")
    bf.add("banana")
    bf.add("cherry")

    # Check membership
    print("apple:", bf.might_contain("apple"))   # Expected: True
    print("banana:", bf.might_contain("banana")) # Expected: True
    print("cherry:", bf.might_contain("cherry")) # Expected: True
    print("grape:", bf.might_contain("grape"))   # Expected: False (or True due to false positive)

    # Debug: Print bit array
    print("Bit Array Snapshot:", bf)


"""
Mathematical Notes:
-------------------
Optimal number of hash functions (k):
    k = (m/n) * ln(2)
False positive probability (FPP):
    FPP ≈ (1 - e^(-kn/m))^k

Where:
- m = size of bit array
- n = number of inserted elements
- k = number of hash functions

Variations:
-----------
1. Counting Bloom Filter → allows deletion using counters instead of bits.
2. Scalable Bloom Filter → grows dynamically for large datasets.
3. Partitioned Bloom Filter → divides bit array for more uniform distribution.

Interview Tip:
--------------
- Be ready to explain *why false negatives are impossible*.
- If asked about tuning, discuss how to pick `m` and `k` given `n` and desired false positive rate.
"""

File: 02_Data_Structures/advanced_structures/bloom_filter/implementations/bloom_filter.cpp
/*
 * Bloom Filter Implementation in C++
 *
 * Problem Statement:
 * ------------------
 * Implement a Bloom Filter – a probabilistic data structure used to test
 * whether an element is a member of a set.
 * False positives are possible, but false negatives are not.
 *
 * Use Case:
 * ---------
 * Common in caching, databases, networking, and security systems where
 * space efficiency and speed are more important than perfect accuracy.
 *
 * Core Operations:
 * ----------------
 * - insert(item): Add an item into the Bloom Filter.
 * - possiblyContains(item): Check if an item might be present in the set.
 *
 * Key Properties:
 * ---------------
 * - False positives possible (may say element exists when it doesn't).
 * - False negatives impossible (never misses actual inserted items).
 *
 * Complexity:
 * -----------
 * - Insertion: O(k)  [k = number of hash functions]
 * - Query: O(k)
 * - Space: O(m) bits [m = size of bit array]
 *
 * Trade-offs:
 * -----------
 * - Increasing m (bit array size) reduces false positive rate.
 * - Increasing k (number of hash functions) reduces false positives up to a point,
 *   but too many increases computation overhead.
 */

#include <iostream>
#include <vector>
#include <bitset>
#include <string>
#include <functional>

class BloomFilter {
private:
    std::vector<bool> bitArray;   // Bit array for Bloom Filter
    int m;                        // Size of bit array
    int k;                        // Number of hash functions
    std::vector<std::hash<std::string>> hashFunctions;

    // Helper: Generate multiple hash values
    size_t hash(const std::string &item, int seed) const {
        std::hash<std::string> hasher;
        return (hasher(item) + seed * 0x9e3779b9) % m;
        // Mixing with seed to generate multiple hash values
    }

public:
    // Constructor
    BloomFilter(int size, int numHashes) : m(size), k(numHashes) {
        bitArray.resize(m, false);
    }

    // Insert an item into the Bloom Filter
    void insert(const std::string &item) {
        for (int i = 0; i < k; i++) {
            int index = hash(item, i);
            bitArray[index] = true;
        }
    }

    // Check if item is possibly present
    bool possiblyContains(const std::string &item) const {
        for (int i = 0; i < k; i++) {
            int index = hash(item, i);
            if (!bitArray[index]) {
                return false;  // Definitely not in set
            }
        }
        return true;  // Might be in set
    }

    // Debug: Print bit array (for visualization)
    void printBitArray() const {
        for (bool bit : bitArray) {
            std::cout << (bit ? "1" : "0");
        }
        std::cout << std::endl;
    }
};

// Driver code
int main() {
    // Example: Bloom Filter with 100 bits and 3 hash functions
    BloomFilter bf(100, 3);

    // Insert elements
    bf.insert("apple");
    bf.insert("banana");
    bf.insert("cherry");

    // Lookup tests
    std::cout << "apple: " << bf.possiblyContains("apple") << std::endl;   // Expected: 1
    std::cout << "banana: " << bf.possiblyContains("banana") << std::endl; // Expected: 1
    std::cout << "cherry: " << bf.possiblyContains("cherry") << std::endl; // Expected: 1
    std::cout << "grape: " << bf.possiblyContains("grape") << std::endl;   // Expected: maybe 0 or 1 (false positive possible)

    // Print bit array (debugging visualization)
    bf.printBitArray();

    return 0;
}

/*
 * Variations / Extensions:
 * ------------------------
 * 1. Counting Bloom Filter → allows deletion by using counters instead of bits.
 * 2. Scalable Bloom Filter → dynamically grows to handle more elements.
 * 3. Partitioned Bloom Filter → divides bit array among hash functions for better distribution.
 *
 * Interview Tip:
 * --------------
 * - If asked in FAANG/system design: explain *why Bloom Filter is space-efficient*
 *   and how false positives are handled.
 * - Mention tuning trade-offs: choose `m` and `k` based on expected n and desired false positive probability:
 *
 *   k = (m/n) * ln(2)
 *   FPP ≈ (1 - e^(-kn/m))^k
 */
